accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+# Link npm packages to the originating source code repository and build
+
+**Authors**: @feelepxyz @kommendorkapten @trevrosen
+
+## Summary
+Link public npm packages to the source code repository and build it originated from, making it harder to execute package hijacking attacks where a malicious version of an existing open source package gets uploaded to the registry.
+
+Check out the [glossary](#glossary) for an overview of the different tools, terms and techniques covered in this RFC document.
+
+## Motivation
+So far our supply chain security offerings have focused on remediation - patching known vulnerabilities in dependencies (e.g. `npm audit`). The majority of vulnerabilities are accidental, so that was the right place to start. Now we need to turn our attention to deliberate supply chain attacks. In 2021, we saw these types of attacks increase by a terrifying [650%](https://www.sonatype.com/resources/state-of-the-software-supply-chain-2021).
+
+In November of last year alone, we saw three supply chain attacks against popular npm packages ([UA-Parser-JS](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack), [COA](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack), and [RC](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack)), each with millions of downloads per month.
+
+The following chart illustrates several attack vectors facing the supply chain:
+
+![supply-chain-threats](https://user-images.githubusercontent.com/20165/174750708-2be483ac-7e41-4bc3-8ee9-440ef33d9423.svg)
+
+Our focus is on mitigating package hijacking attacks where packages are built from a modified source (C) in the above diagram. This is where a malicious version of an existing open source package gets uploaded to the registry. These maliciously crafted versions are rarely reflected in the canonical source code repository and uploaded directly from an attacker's machine.
+
+The [above](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack) [attacks](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack) [are](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack) examples of this, and they frequently occur due to compromised npm credentials but can also happen due to compromised CI/CD or builds.
+
+Code on GitHub, GitLab etc can be browsed and audited, but packages in the registry are opaque and much harder to vet. If packages are built and published in the open these attacks become a whole lot harder. However, right now there’s no way of knowing where a package came from when you retrieve it from the registry.
```

Comment:
unpkg.com makes it much easier - it seems like maybe addressing this (making npm packages browseable on the web safely) would be a good first step?

I'm reasonably certain that the owner of unpkg.com would be happy to transfer the IP to npm if it meant it'd be an official solution (at the least, it's worth asking)

Reply:
We already have a beta implementation of exploring code in npm and are funding improvements to it for live code audit on npmjs.com in parallel to this work.

Reply:
cc @feross